WIP distsql, ui: add statement execution latency graph

rjnn opened the pull request:
Add a graph that tracks the backend execution latency for each SQL statement, which separately tracks DistSQL queries and SQL queries.

WIP as I'm not sure if I'm using the right metric types and setting them up in the nodeGraphs UI page the correct way, and need a quick sanity check.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12790)
<!-- Reviewable:end -->

arjunravinarayan commented:
cc @knz to sanity check that I'm measuring the right thing.

mrtracy commented on pkg/sql/executor.go:
```diff
@@ -56,22 +57,25 @@ var errNotRetriable = errors.New("the transaction is not in a retriable state")
+	MetaTxnRollback        = metric.Metadata{Name: "sql.txn.rollback.count"}
+	MetaSelect             = metric.Metadata{Name: "sql.select.count"}
+	MetaDistSQLSelect      = metric.Metadata{Name: "sql.distsql.select.count"}
+	MetaDistSQLExecLatency = metric.Metadata{Name: "sql.distsql.exec.latency"}
```
These have been added in between the the two stats they are mimicking, can you move this pair below MetaSQLExecLatency?

arjunravinarayan replied:
Done.

mrtracy commented on pkg/sql/executor.go:
```diff
@@ -1341,8 +1354,12 @@ func (e *Executor) execStmt(
 			e.DistSQLSelectCount.Inc(1)
 		}
 		err = e.execDistSQL(planMaker, plan, &result)
+		execDuration := timeutil.Since(tStart).Nanoseconds() + 1000
```
Why the magic number?

arjunravinarayan replied:
Shouldn't be there. Removed.

mrtracy commented on pkg/ui/app/containers/nodeGraphs.tsx:
```diff
@@ -151,6 +151,30 @@ class NodeGraphs extends React.Component<NodeGraphsProps, {}> {
+                       tooltip={`The latency of backend statements executed over 10 second periods.
+                                 Percentiles are calculated on each node.`}
+                       sources={sources}>
+              <Axis format={ (n: number) => d3.format(".1f")(NanoToMilli(n)) } label="Milliseconds">
```
Don't use the "format" or "label" options anymore (I should have removed them); instead, use `<Axis units={AxisUnits.Duration}>`, that does a much better job.

mrtracy commented on pkg/ui/app/containers/nodeGraphs.tsx:
```diff
@@ -151,6 +151,30 @@ class NodeGraphs extends React.Component<NodeGraphsProps, {}> {
+            <LineGraph title="Backend Statement Execution Latency"
+                       subtitle="(Max Per Percentile)"
+                       tooltip={`The latency of backend statements executed over 10 second periods.
+                                 Percentiles are calculated on each node.`}
```
You'll notice the other tooltips have a substitution token `${specifier}`, which gets translated into either "across all node" or "on node X" depending on the currently selected view.  That should be present somewhere in the tooltip.

mrtracy commented:
LGTM with some small changes on the LineGraph and Axis.  You are using the histogram metric correctly and your selected metrics are correct, no issues there.

arjunravinarayan commented:
Thank you for the rapid review, @mrtracy!

Commit:
distsql, ui: add statement execution latency graph

Add a graph that tracks the backend execution latency for each SQL
statement, which separately tracks DistSQL queries and SQL queries.

knz commented:
I am *so* sorry I let this slip under the radar without looking earlier.

This looks good, three questions however:

- what do you intend to happen with implicit txn retries?
- is it all right to log failing and succeeding queries against the same counter?
- your code is (if I understand it correctly) measuring the execution latency, but does not measure the time between the moment the query arrives on pgwire and the moment the last result leaves the server towards the client. I think this last part in particular will be much larger because we are (still) storing the query results in memory during execution and sending them (serially) over the wire after execution completes. What do you make of this? (In other words: why not instrumenting sql/pgwire instead?)

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 1391 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12790#-KcJzyLcEpwI18mmZDFJ:-KcJzyLcEpwI18mmZDFK:b-69mpgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1a5d6266cdeb5af24806358dd94ab364e870e45/pkg/sql/executor.go#L1391)):*
> ```Go
> 		err = e.execClassic(planMaker, plan, &result)
> 		execDuration := timeutil.Since(tStart).Nanoseconds()
> 		e.SQLExecLatency.RecordValue(execDuration)
> ```

Not sure we want to record latencies of failing statement in the same stats as successful statements. What do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12790#-:-KcK-5nd822spWBbg9HQ:b-d68u31)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 1391 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12790#-KcJzyLcEpwI18mmZDFJ:-KcK07eW8KcZWY1e2TUi:biqlgjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1a5d6266cdeb5af24806358dd94ab364e870e45/pkg/sql/executor.go#L1391)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Not sure we want to record latencies of failing statement in the same stats as successful statements. What do you think?
</blockquote></details>

I think we should be consistent with existing SQL latency graphs. (I'm not sure what they do.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12790)*
<!-- Sent from Reviewable.io -->

knz commented:
Well if for any reason the UI currently merges successful and failing queries in one graph, I think that's a UX bug that needs to be fixed :-)

arjunravinarayan commented:
No problem on letting this slip; it was bottlenecked on some UI issues until this Friday, and I had switched to working on other things in the meanwhile.

There is already a latency graph measuring total end-to-end latency. However, the end-to-end system does not know whether it is processing using DistSQL or not using DistSQL. Thus, I did not want to add a DistSQL graph measuring execution latency, while leaving the non-DistSQL graph measuring end-to-end latency (then we are comparing apples to oranges). Thus, I added two graphs, so that DistSQL vs non-DistSQL can be done apples-to-apples execution latency basis, and if someone cares about end-to-end latency, they can look at that (existing) graph as well.

---

Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/executor.go, line 1391 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12790#-KcJzyLcEpwI18mmZDFJ:-KcKWyUab51xAL5tSA5_:b-a4a74l) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1a5d6266cdeb5af24806358dd94ab364e870e45/pkg/sql/executor.go#L1391)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary><blockquote>

I think we should be consistent with existing SQL latency graphs. (I'm not sure what they do.)
</blockquote></details>

The existing SQL latency graphs don't make any distinction, and I'd rather not add 8 graphs with this change. So let's leave it for now, it is an easy add if anyone ever needs them.

---

*[pkg/ui/app/containers/nodeGraphs.tsx, line 157 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12790#-Ka3udhSjo2s2MLuc6BI-r1-157:-KcKJZzX4Sc_KR6tf75u:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaabd43f7c6664d7d6da91c6cb1a94f76655073b/pkg/ui/app/containers/nodeGraphs.tsx#L157)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

You'll notice the other tooltips have a substitution token `${specifier}`, which gets translated into either "across all node" or "on node X" depending on the currently selected view.  That should be present somewhere in the tooltip.
</blockquote></details>

Done.

---

*[pkg/ui/app/containers/nodeGraphs.tsx, line 159 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12790#-Ka3udhSjo2s2MLuc6BI-r1-159:-KcKJ_c7oj9ZwSLgSIHk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aaabd43f7c6664d7d6da91c6cb1a94f76655073b/pkg/ui/app/containers/nodeGraphs.tsx#L159)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Don't use the "format" or "label" options anymore (I should have removed them); instead, use `<Axis units={AxisUnits.Duration}>`, that does a much better job.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12790#-:-KcKH60gWkDpRvKjA_ka:bshyq25)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 1391 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12790#-KcJzyLcEpwI18mmZDFJ:-KcOEEC8W7j_nBHgsen1:bbrs83l) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1a5d6266cdeb5af24806358dd94ab364e870e45/pkg/sql/executor.go#L1391)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

The existing SQL latency graphs don't make any distinction, and I'd rather not add 8 graphs with this change. So let's leave it for now, it is an easy add if anyone ever needs them.
</blockquote></details>

@andreimatei, @raduberinde, do you have an opinion on which latencies would be most useful for you? Do you have an opinion on whether we should log transaction retries, or discard those results?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12790)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/executor.go, line 1391 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12790#-KcJzyLcEpwI18mmZDFJ:-KcUkPlxvi3UIQsSgMB0:bhiblpy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1a5d6266cdeb5af24806358dd94ab364e870e45/pkg/sql/executor.go#L1391)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

@andreimatei, @raduberinde, do you have an opinion on which latencies would be most useful for you? Do you have an opinion on whether we should log transaction retries, or discard those results?
</blockquote></details>

I think this is ok for now. We will need more stats later but they will require extra plumbing in and to DistSQL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12790)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Graphs (low size in order to fit them all in a single screenshot):
![Screenshot 2017-02-21 15.13.05.png](https://files.reviewableusercontent.io/images/reviewable/1565833/93pRHk9pNdOJUXTGAMWL0PrH/Screenshot%202017-02-21%2015.13.05.png)

Would be good to merge before Go 1.8.

---

Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12790#-:-KdXIa-V_uY8TB7NqUlm:b-px4y4s)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Talked to @cuongdo in person, and decided to merge this before Go 1.8 lands.

---

Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12790#-:-KdXL4xMwxzXUHK0DNsX:b-hw3d7m)*
<!-- Sent from Reviewable.io -->
